### PR TITLE
opts: implemet functional parameters the right way

### PR DESCRIPTION
**Pull request by jenstopp**
Signed-off-by: Jens Drenhaus <jens.drenhaus@9elements.com>

### CODE REVIEW

**Comment by jensdrenhaus** [comment]
WIP: 
Adaption in stboot.go needs to be made to use the reworked API of opts pkg.

File: opts/certs_test.go [hunk 1]
```diff
@@ -1,59 +1,52 @@
...
+// 		},
+// 	} {
+// 		t.Run(tt.name, func(t *testing.T) {
+// 			_, err := decodePem(tt.file)
+// 			if err != nil && !errors.Is(err, tt.wantErr) {
+// 				t.Errorf("decodePem(%s)= cert, %q, Want cert, %v", tt.file, err, tt.wantErr)
+// 			}
+// 		})
+// 	}
+// }
```
Comment (RiSKeD) [inline opts/certs_test.go:26]: Why comment out this test and not remove it?

File: opts/certs_test.go [hunk 2]
```diff
@@ -1,59 +1,52 @@
...
+// 			}
+// 		})
+// 		t.Run(tt.name, func(t *testing.T) {
+// 			certLoader := &SigningRootFile{tt.file}
+// 			if _, err := NewOpts(certLoader); err != nil {
+// 				t.Error(err)
+// 			}
+// 		})
+// 	}
+// }
```
Comment (RiSKeD) [inline opts/certs_test.go:52]: Why is this test needed as well?

File: opts/certs.go [hunk 3]
```diff
@@ -2,106 +2,10 @@ package opts
...
-	}
-
-	if len(roots) == 0 {
-		return nil, ErrNoCertificateFound
-	}
-
-	return roots, nil
-}
-
 func printLogCerts(certs ...*x509.Certificate) {
```
Comment (RiSKeD) [inline opts/certs.go:9]: This function is not used anymore

File: opts/certs_test.go (hunk 1 above)
Comment (jensdrenhaus) [reply opts/certs_test.go]: Done

File: opts/certs_test.go (hunk 2 above)
Comment (jensdrenhaus) [reply opts/certs_test.go]: Gone.

**Comment by jensdrenhaus** [comment]
Using `sterror` needs to be done. Maybe separate PR.

**Review by RiSKeD — DISMISSED** [review]
Looks good to me, nicely refactored.

File: opts/opts_test.go [hunk 4]
```diff
@@ -54,3 +47,321 @@ func TestNewOpts(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	keyPEM, err := os.ReadFile("testdata/key.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	badPEM, err := os.ReadFile("testdata/cert_bad_pem.pem")
```
Comment (Foxboron) [inline opts/opts_test.go:196]: Isn't it possible to replace all the calls with `os.Open` which returns an `io.Reader`? Do we strictly need the bytes buffer?

Alternative you could do like this 
```go
func readFile(s string) io.Reader {
	f, err := os.ReadFile(s)
	if err != nil {
		t.Fatal(err)
	}
	return bytes.NewBuffer(f)
}

....

		{
			name:       "Successful loading",
			reader:     readFile(goodCertPEM),
			wantLoaded: true,
			errType:    nil,
		},
```
Or something similar. Prevents the duplicate ReadFile segments and I find it a bit cleaner when extending test cases in the future.

**Comment by Foxboron** [comment]
One small nitpick, rest looks good :)

File: opts/opts_test.go (hunk 4 above)
Comment (jensdrenhaus) [reply opts/opts_test.go:196]: Good one :+1:

**Comment by jensdrenhaus** [comment]
https://git.glasklar.is/system-transparency/core/stboot/-/tree/main/

**Pull request closed without merging**